### PR TITLE
Update Go version

### LINKED ISSUE #310: Update docker image to latest go version
The Docker image is currently using a fixed go v1.20 for the builds.

https://github.com/SchwarzIT/go-template/blob/1152abdc6c4375017591edfc24bbfea0d7e27557/_template/Dockerfile#L4

We should probably update that to use `golang:alpine` without a fixed hash, to always build with the latest Go version. This would bring important security updates, and we wouldn't need to change the version on every new Go release.

**blocked by #313**

### PR DESCRIPTION
**Pull request by zalgonoise**
This PR bumps the Go version in the repository to Go 1.21, and updates its dependencies.

~In order not to impose a specific patch version to the entire project, the toolchain directive is used:~

> `make generate` panics with gomod2nix generation, removed the toolchain directive

This *does not* apply to created projects -- currently the logic picks up the version from a `go version` execution, so created projects would still include that version (1.21.3 in that case).

As for the newly created projects, the minimum version has been bumped to Go 1.20, which is also mentioned on the README file.

Closes #310 #313

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ nix run github:schwarzit/go-template
 
 `go/template`'s `gt` CLI requires at least the following executables on `$PATH` to run succesfully:
 
-- Go >= 1.15
+- Go >= 1.20
```
Comment (MarvinJWendt) [inline README.md:63]:
```suggestion
- Go >= 1.21
```

The go mod is at 1.21 in this PR, so it's required to have Go 1.21 to build the binaries.

File: pkg/gotemplate/new.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ import (
 )
 
 const (
-	minGoVersion  = "1.15"
+	minGoVersion  = "1.20"
```
Comment (MarvinJWendt) [inline pkg/gotemplate/new.go:25]:
```suggestion
	minGoVersion  = "1.21"
```

This should probably also be 1.21, as we want to implement slog (#311), which is only available in Go >= 1.21

**Review by MarvinJWendt — CHANGES_REQUESTED** [review]
Just added some suggestions to set the version to 1.21 instead of 1.20

**Comment by MarvinJWendt** [comment]
Oh, we also have this in the README:

image

Go 1.16+ provides the feature to run `go install`, but Go < 1.21 will not compile the project, so we should probably also bump that version, as using 1.16, 1.17, etc. will result in compiler errors.

File: README.md (hunk 1 above)
Comment (zalgonoise) [reply README.md]: Done and thanks! :D I did it directly when Ctrl+F for occurrences, sorry I didn't take your suggestion here :)

File: pkg/gotemplate/new.go (hunk 2 above)
Comment (zalgonoise) [reply pkg/gotemplate/new.go]: Thank you :) done in here as well

**Comment by zalgonoise** [comment]
> Oh, we also have this in the README:
> > ...

Good catch :D I updated that reference as well

**Comment by zalgonoise** [comment]
Note: I also updated the Dockerfile versions and hashes, as it didn't make much sense to leave those in v1.20 :D 

The distroless-nonroot image [can still be referenced with the same hash](https://console.cloud.google.com/gcr/images/distroless/global/static@sha256:91ca4720011393f4d4cab3a01fa5814ee2714b7d40e6c74f2505f74168398ca9/details), but for the Go image, [we're now referencing the 1.21.3 one](https://hub.docker.com/layers/library/golang/1.21.3/images/sha256-d0214956a9c50c300e430c1f6c0a820007ace238e5242c53762e61b344659e05?context=explore)

Also tagged #310 to close with this PR

**Review by MarvinJWendt — APPROVED** [review]
LGTM! 🚀

File: README.md (hunk 1 above)
Comment (MarvinJWendt) [reply README.md]: No need to be sorry, I only use the suggestions because they look cool. I would also rather commit the changes in the IDE. :)

**Review by danielzwink — APPROVED** [review]

**Pull request merged**